cdap-api/src/main/java/co/cask/cdap/api/plugin/EndpointPluginContext.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.plugin;
+
+import co.cask.cdap.api.annotation.Beta;
+
+/**
+ * Provides access to instantiate plugin for plugin endpoint methods.
+ */
+@Beta
+public interface EndpointPluginContext {
+  /**
+   * Instantiates and returns a plugin instance.
+   * @param pluginType plugin type name
+   * @param pluginName plugin name
+   * @param pluginProperties properties for the plugin.
+   *                         The same set of properties will be used to instantiate the plugin
+   *                         instance at execution time
+   * @param <T> type of the plugin class
+   * @return A {@link Class} for the plugin class or {@code null} if no plugin was found
+   */
+  <T> Class<T> newPluginInstance(String pluginType, String pluginName,
```

Comment:
need one that takes a PluginSelector too, otherwise there is no way to choose which artifact should be used.